File: php/tests/TeamsAndOnboardingTest.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Tests;

use Fivexer\SDK\Model\CreateJoinLink;
use Fivexer\SDK\Model\CreateWorkerIdentity;
use Fivexer\SDK\Model\InviteWorkerIdentity;
use Fivexer\SDK\Model\PatchTeam;
use Fivexer\SDK\Model\UpdateWorkerIdentity;
use Fivexer\SDK\Model\UpsertTeam;
use Fivexer\SDK\Model\WorkerSkillAssignment;

/**
 * Teams, QR join links and worker identities — the roster-building surface.
 *
 * These share one hazard the rest of the API does not: several return a URL that is
 * credential-equivalent until consumed. The tests pin the wire shape and the fact those URLs
 * survive parsing, because a dropped inviteUrl leaves an operator with no way to onboard on a
 * deployment whose mailer is unconfigured — the common case, not the edge case.
 */
final class TeamsAndOnboardingTest extends ClientTestCase
{
    private const TEAM = '{"id":"team_1","key":"billing","tag":"team:billing","name":"Billing",'
        . '"description":"Invoices","color":"#5b8def","createdAt":"2026-08-01T09:00:00.000Z"}';

    private const IDENTITY = '{"id":"wid_1","workerId":"agent_1","label":"Ada",'
        . '"email":"ada@example.com","status":"active","hasPin":false,"activatedAt":null,'
        . '"createdAt":"2026-08-01T09:00:00.000Z","revokedAt":null}';

    // ---- teams ----

    public function testCreatingATeamSendsKeyAndNameAndOmitsUnsetFields(): void
    {
        $this->enqueueJson(201, self::TEAM);

        $team = $this->client()->teams()->create(new UpsertTeam('billing', 'Billing'));

        $request = $this->lastRequest();
        self::assertSame('POST', $request->getMethod());
        self::assertSame('/v1/teams', $this->pathOf($request));
        self::assertSame(['key' => 'billing', 'name' => 'Billing'], $this->requestBodyJson($request));
        // The derived tag is what matching sees, so parsing it is load-bearing.
        self::assertSame('team:billing', $team->tag);
    }

    public function testPatchingATeamOmitsAbsentFieldsSoStoredValuesSurvive(): void
    {
        $this->enqueueJson(200, self::TEAM);

        $this->client()->teams()->patch('team_1', (new PatchTeam())->name('Billing EU'));

        $request = $this->lastRequest();
        self::assertSame('PATCH', $request->getMethod());
        self::assertSame(['name' => 'Billing EU'], $this->requestBodyJson($request));
    }

    public function testListingTeamsUnwrapsTheEnvelope(): void
    {
        $this->enqueueJson(200, '{"teams":[' . self::TEAM . ']}');

        $teams = $this->client()->teams()->list();

        self::assertCount(1, $teams);
        self::assertSame('team_1', $teams[0]->id);
    }

    public function testAnAbsentTeamsArrayReadsAsEmpty(): void
    {
        $this->enqueueJson(200, '{}');

        self::assertSame([], $this->client()->teams()->list());
    }

    public function testTeamMembersCarryTheirRoleAndJoinTime(): void
    {
        $this->enqueueJson(200, '{"teamId":"team_1","members":['
            . '{"workerId":"agent_1","role":"lead","addedAt":"2026-08-02T10:00:00Z"}]}');

        $members = $this->client()->teams()->members('team_1');

        self::assertSame('team_1', $members->teamId);
        self::assertSame('lead', $members->members[0]->role);
    }

    public function testSettingMembersReplacesTheRosterWholesale(): void
    {
        // PUT, not PATCH — a worker absent from the list is removed, and the verb is the only
        // warning a caller gets.
        $this->enqueueJson(200, '{"teamId":"team_1","workerIds":["agent_1"]}');

        $roster = $this->client()->teams()->setMembers('team_1', ['agent_1']);

        $request = $this->lastRequest();
        self::assertSame('PUT', $request->getMethod());
        self::assertSame('/v1/teams/team_1/members', $this->pathOf($request));
        self::assertSame(['workerIds' => ['agent_1']], $this->requestBodyJson($request));
        self::assertSame(['agent_1'], $roster->workerIds);
    }

    public function testRemovingATeamToleratesAnEmpty204(): void
    {
        $this->enqueueEmpty(204);

        $this->client()->teams()->remove('team_1');

        self::assertSame('DELETE', $this->lastRequest()->getMethod());
    }

    public function testGettingATeamEncodesTheIdSegment(): void
    {
        $this->enqueueJson(200, self::TEAM);

        $this->client()->teams()->get('team/../admin');

        self::assertSame('/v1/teams/team%2F..%2Fadmin', $this->pathOf($this->lastRequest()));
    }

    // ---- join links ----

    public function testCreatingAJoinLinkReturnsTheUrlOnlyOnce(): void
    {
        $this->enqueueJson(201, '{"link":{"id":"jl_1","label":"Warehouse","teamId":"team_1",'
            . '"tags":["warehouse"],"requiresApproval":true,"maxUses":25,"useCount":0,'
            . '"status":"active","createdAt":"x","expiresAt":"y","revokedAt":null},'
            . '"joinUrl":"https://5xer.com/join/abc"}');

        $result = $this->client()->joinLinks()->create(
            (new CreateJoinLink('Warehouse'))
                ->teamId('team_1')
                ->tags(['warehouse'])
                ->skills([new WorkerSkillAssignment('sk_1', 3)])
                ->requiresApproval(true)
                ->maxUses(25)
        );

        $body = $this->requestBodyJson($this->lastRequest());
        self::assertSame('Warehouse', $body['label']);
        self::assertSame('sk_1', $body['skills'][0]['skillId']);
        self::assertTrue($body['requiresApproval']);
        // A lost joinUrl is re-created, never recovered — losing it in parsing is unrecoverable.
        self::assertSame('https://5xer.com/join/abc', $result->joinUrl);
        self::assertSame(25, $result->link->maxUses);
    }

    public function testAJoinLinkWithNoCapParsesMaxUsesAsNullNotZero(): void
    {
        // null means unlimited; 0 would mean exhausted. Collapsing them inverts the meaning.
        $this->enqueueJson(200, '{"links":[{"id":"jl_1","label":"Open","teamId":null,"tags":[],'
            . '"requiresApproval":false,"maxUses":null,"useCount":7,"status":"active",'
            . '"createdAt":"x","expiresAt":"y","revokedAt":null}]}');

        $links = $this->client()->joinLinks()->list();

        self::assertNull($links[0]->maxUses);
        self::assertSame(7, $links[0]->useCount);
    }

    public function testRevokingAJoinLinkIsADelete(): void
    {
        $this->enqueueEmpty(204);

        $this->client()->joinLinks()->revoke('jl_1');

        $request = $this->lastRequest();
        self::assertSame('DELETE', $request->getMethod());
        self::assertSame('/v1/join-links/jl_1', $this->pathOf($request));
    }

    public function testAnAbsentLinksArrayReadsAsEmpty(): void
    {
        $this->enqueueJson(200, '{}');

        self::assertSame([], $this->client()->joinLinks()->list());
    }

    // ---- worker identities ----

    public function testInvitingAWorkerReportsWhetherTheMailActuallyWentOut(): void
    {
        // mailer_unconfigured is the common deployment state, and the only signal telling an
        // operator they must hand the link over themselves.
        $this->enqueueJson(201, '{"identity":' . self::IDENTITY . ',"workerCreated":true,'
            . '"emailStatus":"mailer_unconfigured","inviteUrl":"https://5xer.com/i?token=t0k"}');

        $result = $this->client()->identities()->invite(
            (new InviteWorkerIdentity('ada@example.com'))->label('Ada')
        );

        $request = $this->lastRequest();
        self::assertSame('/v1/worker-identities/invite', $this->pathOf($request));
        self::assertSame(['email' => 'ada@example.com', 'label' => 'Ada'], $this->requestBodyJson($request));
        self::assertSame('mailer_unconfigured', $result->emailStatus);
        self::assertStringEndsWith('token=t0k', $result->inviteUrl);
        self::assertTrue($result->workerCreated);
    }

    public function testResendingAnInviteNeedsNoBody(): void
    {
        $this->enqueueJson(200, '{"identity":' . self::IDENTITY . ',"workerCreated":false,'
            . '"emailStatus":"sent","inviteUrl":"https://x/i"}');

        $result = $this->client()->identities()->resendInvite('agent_1');

        self::assertSame('/v1/worker-identities/agent_1/invite/resend', $this->pathOf($this->lastRequest()));
        self::assertFalse($result->workerCreated);
    }

    public function testListingIdentitiesNeverExposesAPinOnlyWhetherOneIsSet(): void
    {
        $this->enqueueJson(200, '{"identities":[' . self::IDENTITY . ']}');

        $identities = $this->client()->identities()->list();

        self::assertFalse($identities[0]->hasPin);
    }

    public function testAnAbsentIdentitiesArrayReadsAsEmpty(): void
    {
        $this->enqueueJson(200, '{}');

        self::assertSame([], $this->client()->identities()->list());
    }

    public function testCreatingAnIdentityDirectlyUnwrapsTheEnvelope(): void
    {
        $this->enqueueJson(201, '{"identity":{"id":"wid_1","workerId":"agent_1","label":"Ada",'
            . '"status":"active","hasPin":true}}');

        $identity = $this->client()->identities()->create('agent_1', new CreateWorkerIdentity('Ada', '4821'));

        $request = $this->lastRequest();
        self::assertSame('/v1/workers/agent_1/identity', $this->pathOf($request));
        self::assertSame(['label' => 'Ada', 'pin' => '4821'], $this->requestBodyJson($request));
        // The caller gets the identity itself, not a one-field wrapper around it.
        self::assertTrue($identity->hasPin);
    }

    public function testAnUnsetEmailIsAbsentFromAnIdentityPatch(): void
    {
        $this->enqueueJson(200, '{"identity":' . self::IDENTITY . '}');

        $this->client()->identities()->update('agent_1', (new UpdateWorkerIdentity())->label('Ada L.'));

        self::assertSame(['label' => 'Ada L.'], $this->requestBodyJson($this->lastRequest()));
    }

    public function testClearingAnIdentityEmailSendsAnExplicitNull(): void
    {
        // Absent means "leave it"; null means "erase it". compact() drops nulls, so this is the
        // one input model that cannot go through the shared serialiser.
        $this->enqueueJson(200, '{"identity":' . self::IDENTITY . '}');

        $this->client()->identities()->update('agent_1', (new UpdateWorkerIdentity())->clearEmail());

        $body = $this->requestBodyJson($this->lastRequest());
        self::assertArrayHasKey('email', $body);
        self::assertNull($body['email']);
    }

    public function testSettingAnEmailAfterClearingItWins(): void
    {
        // The flag must not outlive the decision that set it, or a later email() would silently
        // still erase the address.
        $this->enqueueJson(200, '{"identity":' . self::IDENTITY . '}');

        $this->client()->identities()->update(
            'agent_1',
            (new UpdateWorkerIdentity())->clearEmail()->email('new@example.com')
        );

        self::assertSame('new@example.com', $this->requestBodyJson($this->lastRequest())['email']);
    }

    public function testRevokingAnIdentityIsAStatusPatchNotADelete(): void
    {
        $this->enqueueJson(200, '{"identity":' . self::IDENTITY . '}');

        $this->client()->identities()->update(
            'agent_1',
            (new UpdateWorkerIdentity())->status('revoked')->pin('9137')
        );

        $request = $this->lastRequest();
        self::assertSame('PATCH', $request->getMethod());
        self::assertSame(['pin' => '9137', 'status' => 'revoked'], $this->requestBodyJson($request));
    }

    public function testRemovingAnIdentityLeavesTheWorkerItselfAlone(): void
    {
        // DELETE /workers/{id}/identity, not DELETE /workers/{id} — the worker stays routable,
        // they just lose the ability to sign in.
        $this->enqueueEmpty(204);

        $this->client()->identities()->remove('agent_1');

        $request = $this->lastRequest();
        self::assertSame('DELETE', $request->getMethod());
        self::assertSame('/v1/workers/agent_1/identity', $this->pathOf($request));
    }
}

File: java/src/test/java/io/fivexer/sdk/TeamsAndOnboardingTest.java
package io.fivexer.sdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.google.gson.JsonObject;
import io.fivexer.sdk.model.CreateJoinLink;
import io.fivexer.sdk.model.CreateJoinLinkResult;
import io.fivexer.sdk.model.CreateWorkerIdentity;
import io.fivexer.sdk.model.InviteWorkerIdentity;
import io.fivexer.sdk.model.JoinLink;
import io.fivexer.sdk.model.PatchTeam;
import io.fivexer.sdk.model.PublicWorkerIdentity;
import io.fivexer.sdk.model.Team;
import io.fivexer.sdk.model.TeamMembers;
import io.fivexer.sdk.model.TeamRoster;
import io.fivexer.sdk.model.UpdateWorkerIdentity;
import io.fivexer.sdk.model.UpsertTeam;
import io.fivexer.sdk.model.WorkerInviteResult;
import io.fivexer.sdk.model.WorkerSkillAssignment;
import java.util.List;
import okhttp3.mockwebserver.RecordedRequest;
import org.junit.jupiter.api.Test;

/**
 * Teams, QR join links and worker identities — the roster-building surface.
 *
 * <p>These share one hazard the rest of the API does not: several return a URL that is
 * credential-equivalent until it is consumed. The tests pin the wire shape and the fact that
 * those URLs survive parsing, because a dropped {@code inviteUrl} leaves an operator with no way
 * to onboard on a deployment whose mailer is unconfigured — the common case, not the edge case.
 */
class TeamsAndOnboardingTest extends MockServerBase {

    private static final String TEAM = "{\"id\":\"team_1\",\"key\":\"billing\",\"tag\":\"team:billing\","
            + "\"name\":\"Billing\",\"description\":\"Invoices\",\"color\":\"#5b8def\","
            + "\"createdAt\":\"2026-08-01T09:00:00.000Z\"}";

    private static final String IDENTITY = "{\"id\":\"wid_1\",\"workerId\":\"agent_1\",\"label\":\"Ada\","
            + "\"email\":\"ada@example.com\",\"status\":\"active\",\"hasPin\":false,"
            + "\"activatedAt\":null,\"createdAt\":\"2026-08-01T09:00:00.000Z\",\"revokedAt\":null}";

    // ---- teams ----

    @Test
    void creating_a_team_sends_key_and_name_and_omits_unset_fields() throws Exception {
        enqueueJson(201, TEAM);

        Team team = client().teams().create(new UpsertTeam("billing", "Billing"));

        RecordedRequest request = takeRequest();
        assertEquals("POST", request.getMethod());
        assertEquals("/v1/teams", pathOf(request));
        JsonObject body = bodyOf(request);
        assertEquals("billing", body.get("key").getAsString());
        assertFalse(body.has("description"));
        // The derived tag is what matching sees, so parsing it is load-bearing.
        assertEquals("team:billing", team.getTag());
    }

    @Test
    void patching_a_team_omits_absent_fields_so_stored_values_survive() throws Exception {
        enqueueJson(200, TEAM);

        client().teams().patch("team_1", new PatchTeam().name("Billing EU"));

        RecordedRequest request = takeRequest();
        assertEquals("PATCH", request.getMethod());
        JsonObject body = bodyOf(request);
        assertEquals("Billing EU", body.get("name").getAsString());
        assertEquals(1, body.size());
    }

    @Test
    void listing_teams_unwraps_the_envelope() throws Exception {
        enqueueJson(200, "{\"teams\":[" + TEAM + "]}");

        List<Team> teams = client().teams().list();

        assertEquals(1, teams.size());
        assertEquals("team_1", teams.get(0).getId());
    }

    @Test
    void an_absent_teams_array_reads_as_empty_rather_than_null() throws Exception {
        enqueueJson(200, "{}");

        assertTrue(client().teams().list().isEmpty());
    }

    @Test
    void team_members_carry_their_role_and_join_time() throws Exception {
        enqueueJson(200, "{\"teamId\":\"team_1\",\"members\":["
                + "{\"workerId\":\"agent_1\",\"role\":\"lead\",\"addedAt\":\"2026-08-02T10:00:00Z\"}]}");

        TeamMembers members = client().teams().members("team_1");

        assertEquals("team_1", members.getTeamId());
        assertEquals("lead", members.getMembers().get(0).getRole());
    }

    @Test
    void setting_members_replaces_the_roster_wholesale() throws Exception {
        // PUT, not PATCH — a worker absent from the list is removed, and the verb is the only
        // warning a caller gets.
        enqueueJson(200, "{\"teamId\":\"team_1\",\"workerIds\":[\"agent_1\"]}");

        TeamRoster roster = client().teams().setMembers("team_1", List.of("agent_1"));

        RecordedRequest request = takeRequest();
        assertEquals("PUT", request.getMethod());
        assertEquals("/v1/teams/team_1/members", pathOf(request));
        assertEquals("agent_1", bodyOf(request).getAsJsonArray("workerIds").get(0).getAsString());
        assertEquals(List.of("agent_1"), roster.getWorkerIds());
    }

    @Test
    void removing_a_team_tolerates_an_empty_204() throws Exception {
        enqueueEmpty(204);

        client().teams().remove("team_1");

        assertEquals("DELETE", takeRequest().getMethod());
    }

    @Test
    void getting_a_team_by_id_encodes_the_segment() throws Exception {
        enqueueJson(200, TEAM);

        client().teams().get("team/../admin");

        assertEquals("/v1/teams/team%2F..%2Fadmin", pathOf(takeRequest()));
    }

    // ---- join links ----

    @Test
    void creating_a_join_link_returns_the_url_only_once() throws Exception {
        enqueueJson(201, "{\"link\":{\"id\":\"jl_1\",\"label\":\"Warehouse\",\"teamId\":\"team_1\","
                + "\"tags\":[\"warehouse\"],\"requiresApproval\":true,\"maxUses\":25,\"useCount\":0,"
                + "\"status\":\"active\",\"createdAt\":\"x\",\"expiresAt\":\"y\",\"revokedAt\":null},"
                + "\"joinUrl\":\"https://5xer.com/join/abc\"}");

        CreateJoinLinkResult result = client().joinLinks().create(
                new CreateJoinLink("Warehouse")
                        .teamId("team_1")
                        .tags(List.of("warehouse"))
                        .skills(List.of(new WorkerSkillAssignment("sk_1", 3)))
                        .requiresApproval(true)
                        .maxUses(25));

        RecordedRequest request = takeRequest();
        JsonObject body = bodyOf(request);
        assertEquals("Warehouse", body.get("label").getAsString());
        assertEquals("sk_1", body.getAsJsonArray("skills").get(0).getAsJsonObject()
                .get("skillId").getAsString());
        // A lost joinUrl is re-created, never recovered — losing it in parsing is unrecoverable.
        assertEquals("https://5xer.com/join/abc", result.getJoinUrl());
        assertEquals(25, result.getLink().getMaxUses());
    }

    @Test
    void a_join_link_with_no_cap_parses_max_uses_as_null_not_zero() throws Exception {
        // null means unlimited; 0 would mean exhausted. Collapsing them inverts the meaning.
        enqueueJson(200, "{\"links\":[{\"id\":\"jl_1\",\"label\":\"Open\",\"teamId\":null,"
                + "\"tags\":[],\"requiresApproval\":false,\"maxUses\":null,\"useCount\":7,"
                + "\"status\":\"active\",\"createdAt\":\"x\",\"expiresAt\":\"y\",\"revokedAt\":null}]}");

        List<JoinLink> links = client().joinLinks().list();

        assertNull(links.get(0).getMaxUses());
        assertEquals(7, links.get(0).getUseCount());
    }

    @Test
    void revoking_a_join_link_is_a_delete() throws Exception {
        enqueueEmpty(204);

        client().joinLinks().revoke("jl_1");

        RecordedRequest request = takeRequest();
        assertEquals("DELETE", request.getMethod());
        assertEquals("/v1/join-links/jl_1", pathOf(request));
    }

    @Test
    void an_absent_links_array_reads_as_empty() throws Exception {
        enqueueJson(200, "{}");

        assertTrue(client().joinLinks().list().isEmpty());
    }

    // ---- worker identities ----

    @Test
    void inviting_a_worker_reports_whether_the_mail_actually_went_out() throws Exception {
        // mailer_unconfigured is the common deployment state, and the only signal telling an
        // operator they must hand the link over themselves.
        enqueueJson(201, "{\"identity\":" + IDENTITY + ",\"workerCreated\":true,"
                + "\"emailStatus\":\"mailer_unconfigured\",\"inviteUrl\":\"https://5xer.com/i?token=t0k\"}");

        WorkerInviteResult result = client().identities().invite(
                new InviteWorkerIdentity("ada@example.com").label("Ada"));

        RecordedRequest request = takeRequest();
        assertEquals("/v1/worker-identities/invite", pathOf(request));
        assertEquals("ada@example.com", bodyOf(request).get("email").getAsString());
        assertEquals("mailer_unconfigured", result.getEmailStatus());
        assertTrue(result.getInviteUrl().endsWith("token=t0k"));
        assertTrue(result.getWorkerCreated());
    }

    @Test
    void resending_an_invite_needs_no_body() throws Exception {
        enqueueJson(200, "{\"identity\":" + IDENTITY + ",\"workerCreated\":false,"
                + "\"emailStatus\":\"sent\",\"inviteUrl\":\"https://x/i\"}");

        WorkerInviteResult result = client().identities().resendInvite("agent_1");

        assertEquals("/v1/worker-identities/agent_1/invite/resend", pathOf(takeRequest()));
        assertFalse(result.getWorkerCreated());
    }

    @Test
    void listing_identities_never_exposes_a_pin_only_whether_one_is_set() throws Exception {
        enqueueJson(200, "{\"identities\":[" + IDENTITY + "]}");

        List<PublicWorkerIdentity> identities = client().identities().list();

        assertFalse(identities.get(0).getHasPin());
    }

    @Test
    void an_absent_identities_array_reads_as_empty() throws Exception {
        enqueueJson(200, "{}");

        assertTrue(client().identities().list().isEmpty());
    }

    @Test
    void creating_an_identity_directly_unwraps_the_envelope() throws Exception {
        enqueueJson(201, "{\"identity\":{\"id\":\"wid_1\",\"workerId\":\"agent_1\",\"label\":\"Ada\","
                + "\"status\":\"active\",\"hasPin\":true}}");

        PublicWorkerIdentity identity = client().identities().create(
                "agent_1", new CreateWorkerIdentity("Ada", "4821"));

        RecordedRequest request = takeRequest();
        assertEquals("/v1/workers/agent_1/identity", pathOf(request));
        assertEquals("4821", bodyOf(request).get("pin").getAsString());
        // The caller gets the identity itself, not a one-field wrapper around it.
        assertTrue(identity.getHasPin());
    }

    @Test
    void an_unset_email_is_absent_from_an_identity_patch() throws Exception {
        enqueueJson(200, "{\"identity\":" + IDENTITY + "}");

        client().identities().update("agent_1", new UpdateWorkerIdentity().label("Ada L."));

        JsonObject body = bodyOf(takeRequest());
        assertEquals("Ada L.", body.get("label").getAsString());
        assertFalse(body.has("email"));
    }

    @Test
    void clearing_an_identity_email_sends_an_explicit_null() throws Exception {
        // Absent means "leave it"; null means "erase it". Gson drops nulls, so this is the one
        // input model that cannot go through the shared serialiser.
        enqueueJson(200, "{\"identity\":" + IDENTITY + "}");

        client().identities().update("agent_1", new UpdateWorkerIdentity().clearEmail());

        JsonObject body = bodyOf(takeRequest());
        assertTrue(body.has("email"));
        assertTrue(body.get("email").isJsonNull());
    }

    @Test
    void setting_an_email_after_clearing_it_wins() throws Exception {
        // The flag must not outlive the decision that set it, or a later email() would silently
        // still erase the address.
        enqueueJson(200, "{\"identity\":" + IDENTITY + "}");

        client().identities().update("agent_1",
                new UpdateWorkerIdentity().clearEmail().email("new@example.com"));

        JsonObject body = bodyOf(takeRequest());
        assertEquals("new@example.com", body.get("email").getAsString());
    }

    @Test
    void revoking_an_identity_is_a_status_patch_not_a_delete() throws Exception {
        enqueueJson(200, "{\"identity\":" + IDENTITY + "}");

        client().identities().update("agent_1", new UpdateWorkerIdentity().status("revoked").pin("9137"));

        RecordedRequest request = takeRequest();
        assertEquals("PATCH", request.getMethod());
        JsonObject body = bodyOf(request);
        assertEquals("revoked", body.get("status").getAsString());
        assertEquals("9137", body.get("pin").getAsString());
    }

    @Test
    void removing_an_identity_leaves_the_worker_itself_alone() throws Exception {
        // DELETE /workers/{id}/identity, not DELETE /workers/{id} — the worker stays routable,
        // they just lose the ability to sign in.
        enqueueEmpty(204);

        client().identities().remove("agent_1");

        RecordedRequest request = takeRequest();
        assertEquals("DELETE", request.getMethod());
        assertEquals("/v1/workers/agent_1/identity", pathOf(request));
    }
}
